From 075008dcd42adcc6924d048a376e611f9219e52b Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Thu, 6 Jul 2006 05:10:49 +0000
Subject: [PATCH] Modified getSumSquaredErrors method in SimpleRegression to
 always return a non-negative result. JIRA: MATH-85 Reported by Mark Osborn
 Patched by Luc Maisonobe

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/math/trunk@419439 13f79535-47bb-0310-9956-ffa450edef68
---
 .../math/stat/regression/SimpleRegression.java  | 17 ++++++++++++++++-
 .../stat/regression/SimpleRegressionTest.java   | 13 +++++++++++++
 xdocs/changes.xml                               |  4 ++++
 3 files changed, 33 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/math/stat/regression/SimpleRegression.java b/src/java/org/apache/commons/math/stat/regression/SimpleRegression.java
index c105a41aef..d9fa592c06 100644
--- a/src/java/org/apache/commons/math/stat/regression/SimpleRegression.java
+++ b/src/java/org/apache/commons/math/stat/regression/SimpleRegression.java
@@ -236,6 +236,21 @@ public double getSlope() {
      * sum of squared errors</a> (SSE) associated with the regression 
      * model.
      * <p>
+     * The sum is computed using the computational formula
+     * <p>
+     * <code>SSE = SYY - (SXY * SXY / SXX)</code>
+     * <p>
+     * where <code>SYY</code> is the sum of the squared deviations of the y
+     * values about their mean, <code>SXX</code> is similarly defined and
+     * <code>SXY</code> is the sum of the products of x and y mean deviations.
+     * <p>
+     * The sums are accumulated using the updating algorithm referenced in 
+     * {@link #addData}.  
+     * <p>
+     * The return value is constrained to be non-negative - i.e., if due to 
+     * rounding errors the computational formula returns a negative result, 
+     * 0 is returned.
+     * <p>
      * <strong>Preconditions</strong>: <ul>
      * <li>At least two observations (with at least two different x values)
      * must have been added before invoking this method. If this method is 
@@ -246,7 +261,7 @@ public double getSlope() {
      * @return sum of squared errors associated with the regression model
      */
     public double getSumSquaredErrors() {
-        return sumYY - sumXY * sumXY / sumXX;
+        return Math.max(0d, sumYY - sumXY * sumXY / sumXX);
     }
 
     /**
diff --git a/src/test/org/apache/commons/math/stat/regression/SimpleRegressionTest.java b/src/test/org/apache/commons/math/stat/regression/SimpleRegressionTest.java
index a649b10d22..b6aa67c9dd 100644
--- a/src/test/org/apache/commons/math/stat/regression/SimpleRegressionTest.java
+++ b/src/test/org/apache/commons/math/stat/regression/SimpleRegressionTest.java
@@ -237,6 +237,7 @@ public void testPerfect() throws Exception {
         }
         assertEquals(0.0, regression.getSignificance(), 1.0e-5);
         assertTrue(regression.getSlope() > 0.0);
+        assertTrue(regression.getSumSquaredErrors() >= 0.0);
     }
 
     public void testPerfectNegative() throws Exception {
@@ -261,4 +262,16 @@ public void testRandom() throws Exception {
         assertTrue( 0.0 < regression.getSignificance()
                     && regression.getSignificance() < 1.0);       
     }
+    
+    
+    // Jira MATH-85 = Bugzilla 39432
+    public void testSSENonNegative() {
+        double[] y = { 8915.102, 8919.302, 8923.502 };
+        double[] x = { 1.107178495E2, 1.107264895E2, 1.107351295E2 };
+        SimpleRegression reg = new SimpleRegression();
+        for (int i = 0; i < x.length; i++) {
+            reg.addData(x[i], y[i]);
+        }
+        assertTrue(reg.getSumSquaredErrors() >= 0.0);
+    } 
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 8f33730f89..c534f2137e 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -62,6 +62,10 @@ Commons Math Release Notes</title>
       <action dev="psteitz" type="update" issue="MATH-140" due-to="Xiaogang Zhang">
         Added Fast Fourier Transform implementation.
       </action>
+      <action dev="psteitz" type="fix" issue="MATH-85" due-to="Mark Osborn, Luc Maisonobe">
+        Modified getSumSquaredErrors method in SimpleRegression to always
+        return a non-negative result.
+      </action>
     </release>
     <release version="1.1" date="2005-12-17"  
  description="This is a maintenance release containing bug fixes and enhancements.
